source/common/config/delta_subscription_impl.h

```diff
@@ -261,83 +113,50 @@ class DeltaSubscriptionImpl : public Subscription,
-    // Must not be called if waitingForServer() == true.
-    std::string version() const {
-      ASSERT(version_.has_value());
-      return version_.value_or("");
-    }
-
-  private:
-    absl::optional<std::string> version_;
-  };
-
-  // Use these helpers to avoid forgetting to update both at once.
-  void setResourceVersion(const std::string& resource_name, const std::string& resource_version) {
-    resource_versions_[resource_name] = ResourceVersion(resource_version);
-    resource_names_.insert(resource_name);
-  }
-
-  void setResourceWaitingForServer(const std::string& resource_name) {
-    resource_versions_[resource_name] = ResourceVersion();
-    resource_names_.insert(resource_name);
-  }
-
-  void lostInterestInResource(const std::string& resource_name) {
-    resource_versions_.erase(resource_name);
-    resource_names_.erase(resource_name);
-  }
-
   GrpcStream<envoy::api::v2::DeltaDiscoveryRequest, envoy::api::v2::DeltaDiscoveryResponse>
       grpc_stream_;
 
-  // A map from resource name to per-resource version. The keys of this map are exactly the resource
-  // names we are currently interested in. Those in the waitingForServer state currently don't have
-  // any version for that resource: we need to inform the server if we lose interest in them, but we
-  // also need to *not* include them in the initial_resource_versions map upon a reconnect.
-  std::unordered_map<std::string, ResourceVersion> resource_versions_;
-  // The keys of resource_versions_. Only tracked separately because std::map does not provide an
-  // iterator into just its keys, e.g. for use in std::set_difference.
-  // Must be stored sorted to work with std::set_difference.
-  std::set<std::string> resource_names_;
-
   const std::string type_url_;
-  SubscriptionCallbacks* callbacks_{};
-  // The request being built for the next send.
-  envoy::api::v2::DeltaDiscoveryRequest request_;
-  bool paused_{};
 
   // An item in the queue represents a DeltaDiscoveryRequest that must be sent. If an item is not
-  // empty, it is the ACK (nonce + error_detail) to set on that request. See
-  // trySendDiscoveryRequests() for more details.
+  // empty, it is the ACK (nonce + error_detail) to set on that request. An empty entry should
+  // still send a request; it just won't have an ACK.
+  //
+  // More details: DeltaDiscoveryRequest plays two independent roles:
+  // 1) informing the server of what resources we're interested in, and
+  // 2) acknowledging resources the server has sent us.
+  // Each entry in this queue was added for exactly one of those purposes, but since the
+  // subscription interest is tracked separately, in a non-queue way, subscription changes can get
+  // mixed in with an ACK request. In that case, the entry that the subscription change originally
+  // queued up *does* still get sent, just empty and pointless. (TODO(fredlas) we would like to skip
+  // those no-op requests).
```

Comment:
Yeah, I'm still regretting the choice made early in incremental of keeping state-of-the-world's approach to mixing ACK and subscription change. Even just trying to think through all the corner cases and interleavings here is blowing my mind. Do you reckon it's possible we can make this distinction explicit without causing too much implementation churn?

Reply:
I think the `oneof` approach that we've briefly discussed wouldn't disrupt this code much at all.

Can you clarify for me what aspects of xDS need to be involved in the dependency ordering? Is it just the server's resources and the Envoy's ACKs, or do the Envoy's subscription changes also sometimes need to be ordered with its ACKs in a certain way?

If the subscription changes don't need to be ordered with ACKs, then the two mechanisms seem pretty independent from each other. In that case, this mixing wouldn't actually feel very uncomfortable to me at the conceptual protocol level.

As for the specific implementation here, I definitely do think it's hard to follow, but can be cleaned up. These nullopt queue entries are ultimately a hack to allow subscription changes to use the ACK queue as a way to signal that they need to be sent. There is a cleaner way. The queue can be for ACKs *only* - no longer absl::optional - and the DeltaSubscriptionState can have a `bool subscriptionUpdatePending()`. Then, rather than the sending loop's condition being `!ack_queue_.empty()`, it would be `!ack_queue_.empty() || state_->subscriptionUpdatePending()`. What do you think?

(Note, the above paragraph, as well as the current implementation, would be outright wrong if subscription changes do need to be ordered with ACKs).

Reply:
It would be really easy to reason about if we maintained these properties:
* Each message sent by the server has an ACK/NACK reply at some point in the future; it would be reasonable for the server to wait for this. It could issue a bunch of messages and then wait for their ACKS/NACKs to come back.
* Independent of ^^, Envoy will send a series of subscription changes as needed. These relate in no way to the ACK/NACKs and reflect the client intention.

Does it sound reasonable to make that work from a code perspective?

Reply:
Excellent, thank you. I believe I now have something that takes advantage of that independence to keep things clear, while also still being able to mix them into the same protobuf.

(Also, note the no-op expectSendMessage()s deleted from the test! Hooray!)